### PR TITLE
Add a couple of tests to test passing empty `queryParams` to `replaceWith`

### PR DESCRIPTION
**Pull request by boris-petrov**
Which should fail. Check [this](https://github.com/emberjs/ember.js/pull/19971#issuecomment-1074771144).

### CODE REVIEW

File: packages/ember/tests/routing/router_service_test/replaceWith_test.js [hunk 1]
```diff
@@ -136,5 +136,29 @@ moduleFor(
...
+          assert.deepEqual(this.state, ['/child']);
+        });
+    }
+
+    ['@test RouterService#replaceWith with `undefined` in query params'](assert) {
+      assert.expect(1);
+
+      return this.visit('/')
+        .then(() => {
+          return this.routerService.replaceWith('parent.child', { queryParams: undefined });
```
Comment (wagenet) [inline packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: In this case, this is something we could support but it's also not 100% clear to me that we want to. Can you explain more to me about the use case?

**Review by wagenet — CHANGES_REQUESTED** [review]
I've added some questions.

File: packages/ember/tests/routing/router_service_test/replaceWith_test.js [hunk 2]
```diff
@@ -136,5 +136,29 @@ moduleFor(
           assert.deepEqual(this.state, ['/', '/child']);
         });
     }
+
+    ['@test RouterService#replaceWith with empty query params'](assert) {
+      assert.expect(1);
+
+      return this.visit('/')
+        .then(() => {
+          return this.routerService.replaceWith('parent.child', {});
```
[the remark this answers is not in the document]
Comment (boris-petrov) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js]: Sorry, you're right, fixed!

File: packages/ember/tests/routing/router_service_test/replaceWith_test.js (hunk 1 above)
Comment (boris-petrov) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: Well, you can see what I had in my code [here](https://github.com/emberjs/ember.js/pull/19971#issuecomment-1074771144). The idea is that users of that component could set in `model` their own `routeComponents` and (optionally) `queryParameters` and the component would transition to that. Users might *not* add `queryParameters` which would mean that this call:

```ts
this.router.replaceWith(...this.model.routeComponents, { queryParams: this.model.queryParameters })
```

Would become:

```ts
this.router.replaceWith(...this.model.routeComponents, { queryParams: undefined })
```

Which used to work on Ember < 4.3.0 but broke on 4.3.0. Not saying it's good, just saying it used to work. And is useful. :smile:

Comment (wagenet) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: Fair enough! I'll look into it more next week.

Comment (wagenet) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: So I played around with this some and it looks like router.js would also need an update to handle this case correctly. In your case, a simple solution is probably `this.model.queryParameters || {}`. I'm open to more discussion however.

Comment (boris-petrov) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: hm, but as I said, this used to work in the previous Ember version without changes to `router.js`. But I don't know the current code, of course. Also, note that `this.model.queryParameters || {}` also doesn't work - check the first test of the two - it tests exactly that (and fails).

Comment (wagenet) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: Ah, empty query params should work at least. I'll see about fixing that.

Comment (wagenet) [reply packages/ember/tests/routing/router_service_test/replaceWith_test.js:157]: Also, just because something worked before doesn't mean we guarantee that it will keep working, only if it is something that was clearly documented or widely used to the point where it's basically official. Honestly, this whole API is somewhat awkward so it's not actually clear to me what's most correct. I'm happy to hear another opinion on this.

**Comment by boris-petrov** [comment]
I've involuntarily closed this. I've opened it again as it's still relevant I believe.
